_posts/2020/2020-08-10-spring-boot-upgrade-not-allowed.md

```diff
@@ -0,0 +1,124 @@
+```yaml
+spring:
+  main:
+    allow-bean-definition-overriding: true
+```
+You add that change, modify the microservice to use the needed version of Spring Boot and voilà!
+
+After the same procedure with this service it successfully started.
+But it was not the same anymore.
+
+Every single test in that microservice failed.
+And it failed with a very strange reason:
+```json
+{
+  "status": 405,
+  "error": "Method Not Allowed",
+  "message": "Request method 'GET' not supported",
+  "path": "/api/images"
+}
+```
+It's like there was never an endpoint to get images in an *image storage* microservice.
+By looking deeper into the code I found out that what the microservice actually does is connect to another server and store or fetch the images from it.
+A proxy, if you will.
+And that other server behind the proxy needed [forwarded headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Forwarded){:target="_blank"} to function properly.
+Our proxy is built on top of [Zuul](https://github.com/Netflix/zuul){:target="_blank"}, which was sending the needed headers correctly.
+Emphasis on *was*.
+
+## Why does it have to be beans?
+The problem is simple.
+We need forwarded headers.
+We also need Spring Boot version upgrade.
+What do we do?
+
+It's the Bean time!
+```java
+@Bean
+public Filter forwardedHeaderFilter() {
+    return new ForwardedHeaderFilter();
+}
+```
+
+With this bean in configuration class I run the tests again, and they all successfully pass.
+The proxy plays well with the server behind the proxy, but how do other microservices play with the proxy itself?
+
+We have debug environments for such cases, so that you can quickly check if the changes are working as expected, and also run functional API tests while you are at it.
+I go straight into tests and run all of them, fully expecting them to be green after they finish.
+The tests start one by one and each successfully completes.
+Business, as usual.
+This goes on for some time until one test suddenly fails.
+Then another.
+And another.
+And a dozen more.
+All of them are related to uploading and managing the images on the storage.
+
+And now the vicious loop begins.
+Searching for similar issues, trying out the solutions from all the different posts on Github and Stackoverflow, failing to run them successfully, searching for new solutions, etc.
+Nothing helped.
+My team tried to lend me a hand and after every discussion I would go and make a change in the microservice, only to find out that the changes did not solve the problem.
+
+After a certain point I did not believe that we can upgrade our proxy microservice in the first place and felt like giving up on the task altogether (which means putting it back to the backlog).
+This was when Donald told me he can take this problem over.
+
+---
+
+Integration tests, Distributed Tracing and Logging are the terms that we developers tend to ignore or not worry that much since they don't appear in the scene on sunny days.
```

Comment:
Suggestion: "Integration tests", "Distributed tracing" and "Logging" are the terms...